cms-2016-simulated-datasets/code/mcm_store.py

```diff
@@ -0,0 +1,318 @@
+
+        # check if it is empty (then there is no way to get dataset McM dict)
+        if mcm_step_dict_out == '{"results": {}}\n' or mcm_step_dict_out == '{"results": {}}':
+            print(step + "[ERROR] Empty McM dict (get) for {ds} \n with prepid {pd}".format(ds=dataset,pd=prepid),
+                  file=sys.stderr)
+        else:
+            outfile = step_path + "/dict/" + dataset.replace('/', '@') + ".json"
+            with open(outfile, 'w') as dict_file:
+                    dict_file.write(mcm_step_dict_out)
+
+        mcm_step_script = subprocess.run(cmd + "get_test/" + step ,
+                                shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        mcm_step_script_out = str(mcm_step_script.stdout.decode("utf-8"))
+
+        if mcm_step_script_out == '' or mcm_step_script_out[0] == '{':
+            print(step + " [ERROR] Empty McM script (get_test) for {ds}".format(ds=dataset),
+                file=sys.stderr)
+        else:
+            outfile = step_path + "/scripts/" + dataset.replace('/', '@') + ".sh"
+            with open(outfile, 'w') as dict_file:
+                    dict_file.write(mcm_step_script_out)
+
+    ### End new
+
+
+def create(datasets, mcm_dir, eos_dir, ignore_eos_store=False):
+    "Get information from McM about each dataset"
+
+    # create dirs
+    for path in [mcm_dir + "/dict", mcm_dir + "/scripts", mcm_dir + "/chain"]:
+        os.makedirs(path, exist_ok=True)
+
+    # only for datasets with EOS file information
+    if ignore_eos_store:
+        eos_datasets = datasets.copy()
+    else:
+        eos_datasets = check_datasets_in_eos_dir(datasets, eos_dir)
+
+    total = len(eos_datasets)
+    i = 1
+    for dataset in eos_datasets:
+        print("McM Storing ({i}/{N}) {ds}".format(i=i, N=total, ds=dataset))
+        t = threading.Thread(target=mcm_downloader, args=(dataset, mcm_dir))
+        t.start()
+        while threading.activeCount() >= 100 :
+            sleep(0.5)  # run 100 curl commands in parallel
+        i += 1
+
+
+def get_mcm_dict(dataset, mcm_dir):
+    """Return cached McM dictionary of dataset in json format"""
+
+    filepath = mcm_dir + '/dict/' + dataset.replace('/', '@') + '.json'
+    if os.path.exists(filepath) and os.stat(filepath).st_size != 0:
+        with open(filepath, 'r') as filestream:
+            return json.load(filestream)
+    else:
+        print('[ERROR] There is no McM JSON store dict for dataset ' + dataset,
+              file=sys.stderr)
+        return json.loads('{}')
+
+
+def get_prepId_from_das(dataset, das_dir):
+    "get prepid for dataset"
+
+    # get prepid from das/dataset
+    prepid = get_from_deep_json(get_das_store_json(dataset, 'dataset', das_dir), 'prep_id')
+
+    if prepid == None:
+        # try to get from das/mcm:
+        prepid = get_from_deep_json(get_das_store_json(dataset, 'mcm', das_dir), 'prepid')
+        # todo also try different queries from the json. prep_id?
```

Comment:
Perhaps `# TODO` in uppercase or `# FIXME` so that these can be more easily localised afterwards by editors?